File: src/main/java/com/giant/employee/entity/EmployeeProfile.java
package com.giant.employee.entity;

import com.giant.auth.entity.Account;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.time.Instant;

@Entity
@Table(
        name = "profile",
        schema = "employee",
        indexes = {
                @Index(name = "fk_profile_team", columnList = "team_id"),
                @Index(name = "fk_profile_role", columnList = "employee_role_id"),
                @Index(name = "fk_profile_position", columnList = "position_id"),
                @Index(name = "idx_profile_created_at", columnList = "created_at"),
                @Index(name = "idx_profile_updated_at", columnList = "updated_at"),
                @Index(name = "idx_profile_left_at", columnList = "left_at"),
                @Index(name = "idx_profile_created_by", columnList = "created_by"),
                @Index(name = "idx_profile_updated_by", columnList = "updated_by")
        }
)
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class EmployeeProfile {
    @Id
    @Column(name="id")
    private Long employeeId;

    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false, foreignKey = @ForeignKey(name = "fk_profile_account"))
    private Account account;

    @Column(
            name = "employee_code",
            nullable = false,
            unique = true
    )
    private String employeeCode;

    @Column(
            name = "employee_name",
            nullable = false
    )
    private String employeeName;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = false, foreignKey = @ForeignKey(name = "fk_profile_team"))
    private Team team;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_role_id", nullable = false, foreignKey = @ForeignKey(name = "fk_profile_role"))
    private EmployeeRole employeeRole;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "position_id", nullable = false, foreignKey = @ForeignKey(name = "fk_profile_position"))
    private Position position;

    @Column(name = "created_by", nullable = false, updatable = false)
    private Long createdBy;

    @Column(name = "updated_by")
    private Long updatedBy = null;

    @Column(
            name = "created_at",
            nullable = false,
            updatable = false,
            columnDefinition = "TIMESTAMP(3) WITHOUT TIME ZONE DEFAULT (NOW() AT TIME ZONE 'UTC')"
    )
    private Instant createdAt;

    @Column(
            name = "updated_at",
            nullable = false,
            columnDefinition = "TIMESTAMP(3) WITHOUT TIME ZONE DEFAULT (NOW() AT TIME ZONE 'UTC')"
    )
    private Instant updatedAt;

    @Column(name = "left_at", columnDefinition = "TIMESTAMP(3) WITHOUT TIME ZONE")
    private Instant leftAt = null;

    @PrePersist
    protected void onCreate() {
        updatedAt = createdAt = Instant.now();
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = Instant.now();
    }
}

File: src/main/kotlin/com/giant/employee/entity/EmployeeProfile.kt
package com.giant.employee.entity

import com.giant.auth.entity.Account
import jakarta.persistence.*
import org.hibernate.annotations.DynamicUpdate
import java.time.Clock
import java.time.LocalDateTime

@Entity
@Table(
    name = "employee_profile",
    schema = "employee",
    indexes = [
        Index(name = "idx_employee_team_id", columnList = "team_id"),
        Index(name = "idx_employee_role_id", columnList = "employee_role_id"),
        Index(name = "idx_employee_position_id", columnList = "position_id")
    ]
)
@DynamicUpdate
class EmployeeProfile(

    @Id
    @Column(name = "id")
    var id: Long = 0,

    @MapsId
    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "id", nullable = false, foreignKey = ForeignKey(name = "fk_employee_account"))
    var account: Account? = null,

    @Column(
        name = "employee_code",
        nullable = false,
        unique = true,
        length = 255,
        columnDefinition = "nvarchar(255) COLLATE Latin1_General_100_CI_AS_SC"
    )
    var employeeCode: String = "",

    @Column(
        name = "employee_name",
        nullable = false,
        length = 255,
        columnDefinition = "nvarchar(255) COLLATE Latin1_General_100_CI_AS_SC"
    )
    var employeeName: String = "",

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "team_id", nullable = false, foreignKey = ForeignKey(name = "fk_employee_team"))
    var team: Team? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "employee_role_id", nullable = false, foreignKey = ForeignKey(name = "fk_employee_role"))
    var employeeRole: EmployeeRole? = null,

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "position_id", nullable = false, foreignKey = ForeignKey(name = "fk_employee_position"))
    var position: Position? = null,

    @Column(name = "created_by", nullable = false, updatable = false)
    var createdBy: Long = 0,

    @Column(name = "updated_by")
    var updatedBy: Long? = 0,

    @Column(
        name = "created_at",
        nullable = false,
        updatable = false,
        columnDefinition = "datetime2(3) default SYSUTCDATETIME()"
    )
    var createdAt: LocalDateTime? = null,

    @Column(
        name = "updated_at",
        nullable = false,
        columnDefinition = "datetime2(3) default SYSUTCDATETIME()"
    )
    var updatedAt: LocalDateTime? = null,

    @Column(name = "left_at", columnDefinition = "datetime2(3)")
    var leftAt: LocalDateTime? = null
) {
    @PrePersist
    fun prePersist() {
        val now = LocalDateTime.now(Clock.systemUTC())
        createdAt = now
        updatedAt = now
    }

    @PreUpdate
    fun preUpdate() {
        updatedAt = LocalDateTime.now(Clock.systemUTC())
    }
}